diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index f991e525c..1f2dd22a3 100644
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -1893,22 +1893,12 @@ config("default_warnings") {
 
         # TODO(crbug.com/1494809): Evaluate and possibly enable.
         "-Wno-vla-extension",
-
-        # TODO(https://crbug.com/1490607): Fix and re-enable.
-        "-Wno-thread-safety-reference-return",
       ]
 
       cflags_cc += [
         # TODO(crbug.com/328490295): Fix and re-enable for C flags.
         "-Wenum-compare-conditional",
       ]
-
-      if (!is_nacl) {
-        cflags_cc += [
-          # TODO(https://crbug.com/1513724): Fix and re-enable.
-          "-Wno-c++11-narrowing-const-reference",
-        ]
-      }
     }
 
     # Some builders, such as Cronet, use a different version of Clang than
diff --git a/config/ios/BUILD.gn b/config/ios/BUILD.gn
index 44992837c..cbc212683 100644
--- a/config/ios/BUILD.gn
+++ b/config/ios/BUILD.gn
@@ -256,6 +256,14 @@ group("xctest") {
 _xctrunner_path =
     "$ios_sdk_platform_path/Developer/Library/Xcode/Agents/XCTRunner.app"
 
+if (target_environment == "catalyst") {
+  _xctrunner_executable_path =
+      "$_xctrunner_path/Contents/MacOS/XCTRunner"
+} else {
+  _xctrunner_executable_path =
+      "$_xctrunner_path/XCTRunner"
+}
+
 # When building with Goma RBE, $ios_sdk_platform_path corresponds to a symlink
 # below $root_build_dir that points to the real SDK to use. Because the files
 # are below $root_build_dir, it is not possible to list them as a target input
@@ -269,15 +277,24 @@ _xctrunner_path =
 # To workaround this, add a target that pretends to create those files
 # (but does nothing). See https://crbug.com/1061487 for why this is needed.
 if (use_system_xcode && (use_goma || use_remoteexec || use_siso)) {
-  action("copy_xctrunner_app") {
-    testonly = true
-    script = "//build/noop.py"
+  if (target_environment == "catalyst") {
+    outputs = [
+      "$_xctrunner_path/Contents/Info.plist",
+      "$_xctrunner_path/Contents/PkgInfo",
+      "$_xctrunner_executable_path",
+    ]
+  } else {
     outputs = [
       "$_xctrunner_path/Info.plist",
       "$_xctrunner_path/PkgInfo",
-      "$_xctrunner_path/XCTRunner",
+      "$_xctrunner_executable_path",
     ]
   }
+  action("copy_xctrunner_app") {
+    testonly = true
+    script = "//build/noop.py"
+    outputs = outputs
+  }
 }
 
 # When creating the test runner for an XCUITest, the arm64e slice of the binary
@@ -286,7 +303,7 @@ if (use_system_xcode && (use_goma || use_remoteexec || use_siso)) {
 action("xctest_runner_without_arm64e") {
   testonly = true
   script = "//build/config/ios/strip_arm64e.py"
-  sources = [ "$_xctrunner_path/XCTRunner" ]
+  sources = [ _xctrunner_executable_path ]
   outputs = [ "$target_out_dir/XCTRunner" ]
   args = [
     "--output",
diff --git a/config/ios/rules.gni b/config/ios/rules.gni
index 3f47b394e..83e0e346f 100644
--- a/config/ios/rules.gni
+++ b/config/ios/rules.gni
@@ -1805,9 +1805,13 @@ template("ios_xcuitest_test_runner_bundle") {
     testonly = true
     script = "//build/apple/plist_util.py"
 
-    sources = [
-      "$_xctrunner_path/Info.plist",
+    if (target_environment == "catalyst") {
+      sources = ["$_xctrunner_path/Contents/Info.plist"]
+    } else {
+      sources = ["$_xctrunner_path/Info.plist"]
+    }
 
+    sources += [
       # NOTE: The XCTRunnerAddition+Info.plist must come after the Info.plist
       # because it overrides the values under "CFBundleIdentifier" and
       # "CFBundleName".
@@ -1852,7 +1856,11 @@ template("ios_xcuitest_test_runner_bundle") {
     testonly = true
     visibility = [ ":$_target_name" ]
 
-    sources = [ "$_xctrunner_path/PkgInfo" ]
+    if (target_environment == "catalyst") {
+      sources = ["$_xctrunner_path/Contents/PkgInfo"]
+    } else {
+      sources = ["$_xctrunner_path/PkgInfo"]
+    }
 
     outputs = [ "{{bundle_contents_dir}}/PkgInfo" ]
 
