diff --git a/config/apple/sdk_info.py b/config/apple/sdk_info.py
index 81b06d438..94b260971 100755
--- a/config/apple/sdk_info.py
+++ b/config/apple/sdk_info.py
@@ -134,7 +134,7 @@ def main():
                       default='.',
                       help='Value of gn $root_build_dir')
   parser.add_argument('platform',
-                      choices=['iphoneos', 'iphonesimulator', 'macosx'])
+                      choices=['iphoneos', 'iphonesimulator', 'macosx', 'xros', 'xrsimulator'])
   args = parser.parse_args()
   if args.developer_dir:
     os.environ['DEVELOPER_DIR'] = args.developer_dir
diff --git a/config/clang/BUILD.gn b/config/clang/BUILD.gn
index fcca59b0f..1abb3431f 100644
--- a/config/clang/BUILD.gn
+++ b/config/clang/BUILD.gn
@@ -240,7 +240,13 @@ clang_lib("compiler_builtins") {
   if (is_mac) {
     libname = "osx"
   } else if (is_ios) {
-    if (target_environment == "simulator") {
+    if (target_os_variation == "visionos") {
+      if (target_environment == "simulator") {
+        libname = "xrossim"
+      } else {
+        libname = "xros"
+      }
+    } else if (target_environment == "simulator") {
       libname = "iossim"
     } else if (target_environment == "catalyst") {
       libname = "osx"
diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index 750d292a0..486c25e2c 100644
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -586,24 +586,6 @@ config("compiler") {
       }
     }
 
-    # TODO(crbug.com/40283598): This causes binary size growth and potentially
-    # other problems.
-    # TODO(crbug.com/40284925): This isn't supported by Cronet's mainline llvm version.
-    if (default_toolchain != "//build/toolchain/cros:target" &&
-        !llvm_android_mainline) {
-      cflags += [
-        "-mllvm",
-        "-split-threshold-for-reg-with-hint=0",
-      ]
-      if (use_thin_lto && is_a_target_toolchain) {
-        if (is_win) {
-          ldflags += [ "-mllvm:-split-threshold-for-reg-with-hint=0" ]
-        } else {
-          ldflags += [ "-Wl,-mllvm,-split-threshold-for-reg-with-hint=0" ]
-        }
-      }
-    }
-
     # TODO(crbug.com/40192287): Investigate why/if this should be needed.
     if (is_win) {
       cflags += [ "/clang:-ffp-contract=off" ]
@@ -1891,6 +1873,7 @@ config("default_warnings") {
       # Disables.
       "-Wno-missing-field-initializers",  # "struct foo f = {0};"
       "-Wno-unused-parameter",  # Unused function parameters.
+      "-Wno-deprecated-declarations",
     ]
 
     if (!is_nacl || is_nacl_saigo) {
diff --git a/config/ios/BUILD.gn b/config/ios/BUILD.gn
index 42c8ed6ad..b8bc61343 100644
--- a/config/ios/BUILD.gn
+++ b/config/ios/BUILD.gn
@@ -40,7 +40,11 @@ config("compiler") {
   }
 
   # OS.
-  triplet_os = "apple-ios"
+  if (target_os_variation == "visionos") {
+    triplet_os = "apple-xros"
+  } else {
+    triplet_os = "apple-ios"
+  }
 
   # Set target.
   common_flags = [
diff --git a/config/ios/codesign.py b/config/ios/codesign.py
index fd96f312d..b2f43cb9b 100644
--- a/config/ios/codesign.py
+++ b/config/ios/codesign.py
@@ -91,6 +91,8 @@ class Bundle(object):
   def Kind(platform, extension):
     if platform == 'iphonesimulator' or platform == 'iphoneos':
       return 'ios'
+    if platform == 'xros' or platform == 'xrsimulator':
+      return 'xros'
     if platform == 'macosx':
       if extension == '.framework':
         return 'mac_framework'
diff --git a/config/ios/config.gni b/config/ios/config.gni
index 2b7a8f60c..480a87e88 100644
--- a/config/ios/config.gni
+++ b/config/ios/config.gni
@@ -9,6 +9,7 @@ declare_args() {
   # default is only there for compatibility reasons and will be removed (see
   # crbug.com/1138425 for more details).
   target_environment = ""
+  target_os_variation = ""
 
   # Generate orderfile at application startup and then exit.
   # NOTE: This flag adds runtime tooling to capture function call details,
diff --git a/config/ios/ios_sdk.gni b/config/ios/ios_sdk.gni
index fd9f6d175..f8555072d 100644
--- a/config/ios/ios_sdk.gni
+++ b/config/ios/ios_sdk.gni
@@ -97,7 +97,17 @@ ios_use_xcode_symlinks =
 
 if (ios_sdk_path == "") {
   # Compute default target.
-  if (target_environment == "simulator") {
+  if (target_os_variation == "visionos") {
+    if (target_environment == "simulator") {
+      ios_sdk_name = "xrsimulator"
+      ios_sdk_platform = "XRSimulator"
+    } else if (target_environment == "device") {
+      ios_sdk_name = "xros"
+      ios_sdk_platform = "XROS"
+    } else {
+      assert(false, "unsupported environment: $target_environment")
+    }
+  } else if (target_environment == "simulator") {
     ios_sdk_name = "iphonesimulator"
     ios_sdk_platform = "iPhoneSimulator"
   } else if (target_environment == "device") {
diff --git a/toolchain/apple/toolchain.gni b/toolchain/apple/toolchain.gni
index acd5a59f5..5870c5368 100644
--- a/toolchain/apple/toolchain.gni
+++ b/toolchain/apple/toolchain.gni
@@ -449,7 +449,7 @@ template("single_apple_toolchain") {
         command = "rm -f {{output}} && TOOL_VERSION=${tool_versions.filter_libtool} $python_path $script $libtool -static -D {{arflags}} -o {{output}} @$rspfile"
         description = "LIBTOOL-STATIC {{output}}"
       } else {
-        ar = "${prefix}llvm-ar"
+        ar = "../../../third_party/llvm-build/Release+Asserts/bin/llvm-ar"
         command = "\"$ar\" {{arflags}} -r -c -s -D {{output}} @$rspfile"
 
         # Remove the output file first so that ar doesn't try to modify the
